- adds detailed migration guidelines for the authentication providers (#647)
microsoftgraph/msgraph-sdk-java: docs/upgrade-to-v3.md

1. Remove any dependency on `microsoft-graph-android-auth` and replace them by one of the providers available in azure-identity.
1. Reconsider the need for custom authentication providers (implementation of `IAuthenticationProvider` or `ICoreAuthenticationProvider`) that might be part of your project and use a provider available in azure-identity whenever possible.

> #### Authentication upgrade examples

Reviewer: It would be helpful to move these examples into a separate doc say TokenCredentialSamples.md. It would be easier to find this. I think of the upgrade doc as more of a change log.

Author: thanks for the suggestion. updated.

The purpose of this section is to provide examples on how to upgrade your authentication providers to azure-identity providers.

##### Authorization code flow

Replace any of the following:

```Java
AuthorizationCodeProvider authProvider = new AuthorizationCodeProvider(CLIENT_ID, SCOPES, AUTHORIZATION_CODE, REDIRECT_URL, NATIONAL_CLOUD, TENANT, CLIENT_SECRET);
IGraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
```

By

``` Java
final AuthorizationCodeCredential authCodeCredential = new AuthorizationCodeCredentialBuilder()
    .clientId(CLIENT_ID)

>     .clientSecret(CLIENT_SECRET) //required for web apps, do not set for native apps

Reviewer 2: What does "not setting for native apps" look like?
- Should this call be skipped?
- Should they pass in null or empty string?

Author: ```suggestion
    .clientSecret(CLIENT_SECRET) //required for web apps, skip it for native apps
```
would this be better?

Reviewer 2: I think that works.

    .authorizationCode(AUTHORIZATION_CODE)
    .redirectUrl(REDIRECT_URL)
    .build();
final TokenCredentialAuthProvider tokenCredentialAuthProvider = new TokenCredentialAuthProvider(SCOPES, authCodeCredential);

> final GraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();

Reviewer 3: Shouldn't it be "tokenCredentialAuthProvider"?

Author: thanks, fixed.

```

##### Client credential flow

Replace any of the following:

```Java
ClientCredentialProvider authProvider = new ClientCredentialProvider(CLIENT_ID, SCOPES, CLIENT_SECRET, TENANT_GUID, NATIONAL_CLOUD);
IGraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
```

By

``` Java
final ClientSecretCredential clientSecretCredential = new ClientSecretCredentialBuilder()
    .clientId(CLIENT_ID)
    .clientSecret(CLIENT_SECRET)
    .tenantId(TENANT_GUID)
    .build();
final TokenCredentialAuthProvider tokenCredentialAuthProvider = new TokenCredentialAuthProvider(SCOPES, authCodeCredential);

> final GraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();

Reviewer 3: Same here:
authCodeCredential -> clientSecretCredential 
authProvider -> tokenCredentialAuthProvider

Author: thanks, fixed.

```

> Note: you can alternatively use **ClientCertificateCredential** to authenticate using a certificate instead of a secret. Using certificates for authentication was not previously supported.

##### User name and password flow

Replace any of the following:

```Java
UsernamePasswordProvider authProvider = new UsernamePasswordProvider(CLIENT_ID, SCOPES, USERNAME, PASSWORD, NATIONAL_CLOUD, TENANT, CLIENT_SECRET);
IGraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
```

By

``` Java
final UsernamePasswordCredential usernamePasswordCredential = new UsernamePasswordCredentialBuilder()
    .clientId(CLIENT_ID)
    .username(USERNAME)
    .password(PASSWORD)
    .build();
final TokenCredentialAuthProvider tokenCredentialAuthProvider = new TokenCredentialAuthProvider(SCOPES, authCodeCredential);

> final GraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();

Reviewer 3: authCodeCredential -> usernamePasswordCredential 
authProvider -> tokenCredentialAuthProvider

Author: thanks, fixed.

```

##### Device code flow

> Note: the device code flow was previously not available, some customers might have implemented custom authentication providers in their application to bridge the feature gap. That custom code can now be replaced.

Replace any of the following:

```Java
CustomBuiltDeviceCodeProvider authProvider = new CustomBuiltDeviceCodeProvider(CLIENT_ID); // code specific to the custom implementation
IGraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
```

By

``` Java
final DeviceCodeCredential deviceCodeCredential = new DeviceCodeCredentialBuilder()
    .clientId(CLIENT_ID)
    .challengeConsumer(challenge -> {
                        // lets user know of the challenge
                        System.out.println(challenge.getMessage());
                    })
    .build();
final TokenCredentialAuthProvider authProvider = new TokenCredentialAuthProvider(SCOPES, deviceCodeCredential);
final GraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
```

> ##### Interwactive browser flow

Reviewer 2: ```suggestion
##### Interactive browser flow
```

Author: ahhh memory muscle blip from the first part of my life being on AZERTY keyboards....

Reviewer 2: AZERTY! Reminds me of my internship days in France :)

> final GraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();

Reviewer 3: authCodeCredential -> deviceCodeCredential
authProvider -> tokenCredentialAuthProvider

Author: thanks, fixed.

Replace any of the following:

```Java
PublicClientApplication publicClientApplication = new PublicClientApplication(getApplicationContext(), CLIENT_ID);
MSALAuthenticationProvider msalAuthenticationProvider = new MSALAuthenticationProvider(
    getActivity(),
    getApplication(),
    publicClientApplication,
    scopes);
CustomBuiltBrowserProvider authProvider = new CustomBuiltBrowserProvider(CLIENT_ID, REDIRECT_URL);
IGraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
```

And

```Java
@Override
protected void onActivityResult(int requestCode, int resultCode, Intent data) {
    msalAuthenticationProvider.handleInteractiveRequestRedirect(requestCode, resultCode, data);
}
```

By

``` Java
final InteractiveBrowserCredential interactiveBrowserCredential = new InteractiveBrowserCredentialBuilder()
    .clientId(CLIENT_ID)
    .redirectUrl(REDIRECT_URL)
    .build();
final TokenCredentialAuthProvider tokenCredentialAuthProvider = new TokenCredentialAuthProvider(SCOPES, authCodeCredential);

> final GraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();

Reviewer 3: authCodeCredential -> interactiveBrowserCredential 
authProvider -> tokenCredentialAuthProvider

Author: thanks, fixed.
